src/Lean/Meta/Tactic/Simp/SimpAll.lean

```diff
@@ -118,17 +120,26 @@ def main : M (Option MVarId) := do
     return none -- close the goal
   else
     let mvarId := (← get).mvarId
-    let entries := (← get).entries
-    let (_, mvarId) ← mvarId.assertHypotheses <| entries.filterMap fun e =>
-      -- Do not assert `True` hypotheses
-      if e.type.consumeMData.isConstOf ``True then none else some { userName := e.userName, type := e.type, value := e.proof }
-    mvarId.tryClearMany (entries.map fun e => e.fvarId)
+    let mut toAssert := #[]
+    let mut toClear := #[]
+    for e in (← get).entries do
+      if e.type.consumeMData.isConstOf ``True then
+        -- Do not assert `True` hypotheses
+        toClear := toClear.push e.fvarId
+      else if e.type != e.origType then
+        toClear := toClear.push e.fvarId
+        toAssert := toAssert.push { userName := e.userName, type := e.type, value := e.proof }
```

Comment:
> I'm sorry about this, I didn't think the hypothesis reordering would be important. One thing we could do to preserve hypothesis ordering but also preserve the property that if no hypotheses need to change then nothing is done, is to have a flag

> This change must be documented in the change log: https://github.com/leanprover/lean4/blob/master/RELEASES.md We must also document why it has been made, and add links to zulip threads where the discussion happened, and how the consensus has been reached. Reason: this change will affect (non mathlib) users.

Sorry, this is partially my fault, I didn't think the hypothesis reordering would be important. One thing we could do to preserve hypothesis ordering but also preserve the property that if no hypotheses need to change then nothing is done, is to have a flag which is set when a prior hypothesis has been rewritten and then just rewrite all subsequent hypotheses, like so:

```suggestion
    let mut modified := false
    for e in (← get).entries do
      if e.type.consumeMData.isConstOf ``True then
        -- Do not assert `True` hypotheses
        toClear := toClear.push e.fvarId
      else if modified || e.type != e.origType then
        toClear := toClear.push e.fvarId
        toAssert := toAssert.push { userName := e.userName, type := e.type, value := e.proof }
        modified := true
```
Here's a patch that will

Reply:
Okay, I've pushed this and cleaned up.